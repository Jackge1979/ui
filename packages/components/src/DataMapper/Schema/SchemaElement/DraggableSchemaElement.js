import React, { Component } from 'react';
import { ItemTypes, SchemaType } from '../../Constants';
import { DragSource } from 'react-dnd';
import { DropTarget } from 'react-dnd';
import PropTypes from 'prop-types';
import flow from 'lodash/flow';
import SchemaElement from './SchemaElement.js';

const elementSource = {
	beginDrag(props) {
		return {
			elementId: props.name,
			schemaType: props.schemaType,
		};
	},
};

const elementTarget = {
	drop(props, monitor) {
		var sourceElem = monitor.getItem();
		props.performMapping(sourceElem.elementId, props.name);
	},
	canDrop(props, monitor) {
		var sourceElem = monitor.getItem();
		return props.schemaType !== sourceElem.schemaType && sourceElem.schemaType === SchemaType.INPUT;
	},
};

function collectForDragSource(connect, monitor) {
	return {
		connectDragSource: connect.dragSource(),
		isDragging: monitor.isDragging(),
	};
}

function collectForDropTarget(connect, monitor) {
	return {
		connectDropTarget: connect.dropTarget(),
		dragOver: monitor.isOver(),
	};
}

class DraggableSchemaElement extends Component {
	render() {
		const {
			name,
			schemaType,
			mapped,
			connectDragSource,
			connectDropTarget,
			onSelect,
			selected,
		} = this.props;
		const highlighted = this.props.dragOver;
		return connectDragSource(
			connectDropTarget(
				<div>
					<SchemaElement
						highlighted={highlighted}
						mapped={mapped}
						name={name}
						schemaType={schemaType}
						selected={selected}
						onSelect={onSelect}
					/>
				</div>,
			),
		);
	}
}

DraggableSchemaElement.propTypes = {
	name: PropTypes.string,
	schemaType: PropTypes.string,
};

export default flow(
	DragSource(ItemTypes.ELEMENT, elementSource, collectForDragSource),
	DropTarget(ItemTypes.ELEMENT, elementTarget, collectForDropTarget),
)(DraggableSchemaElement);
