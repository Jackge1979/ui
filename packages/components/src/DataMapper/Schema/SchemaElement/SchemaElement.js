import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Returns the className of the schema element depending of its props.
 * @param {object} props - The props of the schema element.
 */
function getClassName(props) {
	let className = 'schema-element';
	if (props.highlighted) {
		className += ' highlighted';
	}
	if (props.mapped) {
		className += ' mapped';
	}
	if (props.selected) {
		className += ' selected';
	}
	className += ` ${props.side}`;
	return className;
}

/**
 * A single element of a schema. It is mainly a name with a side (i/o).
 * @author timbault
 */
export default class SchemaElement extends Component {

	constructor(props) {
		super(props);
		this.select = this.select.bind(this);
		this.revealConnection = this.revealConnection.bind(this);
		this.handleMouseEnter = this.handleMouseEnter.bind(this);
		this.handleMouseLeave = this.handleMouseLeave.bind(this);
		this.updateElementRef = this.updateElementRef.bind(this);
	}

	componentDidMount() {
		if (this.elementRef != null) {
			this.elementRef.addEventListener('mouseenter', this.handleMouseEnter);
			this.elementRef.addEventListener('mouseleave', this.handleMouseLeave);
		}
	}

	componentWillUnmount() {
		if (this.elementRef != null) {
			this.elementRef.removeEventListener('mouseenter', this.handleMouseEnter);
			this.elementRef.removeEventListener('mouseleave', this.handleMouseLeave);
		}
	}

	handleMouseEnter() {
		this.props.onEnterElement(this.props.element, this.props.side);
	}

	handleMouseLeave() {
		this.props.onLeaveElement(this.props.element, this.props.side);
	}

	select(ev) {
		this.props.onSelect(ev.ctrlKey, this.props.element, this.props.side);
	}

	revealConnection() {
		this.props.revealConnection(this.props.element, this.props.side);
	}

	updateElementRef(ref) {
		this.elementRef = ref;
	}

	render() {
		return (
			<div
				className={getClassName(this.props)}
				onClick={this.select}
				onDoubleClick={this.revealConnection}
				role="button"
				ref={this.updateElementRef}
			>
				{this.props.dataAccessor.getElementName(this.props.element)}
			</div>
		);
	}
}

SchemaElement.propTypes = {
	dataAccessor: PropTypes.object,
	element: PropTypes.any.isRequired,
	side: PropTypes.string,
	onSelect: PropTypes.func,
	onEnterElement: PropTypes.func,
	onLeaveElement: PropTypes.func,
	revealConnection: PropTypes.func,
};
