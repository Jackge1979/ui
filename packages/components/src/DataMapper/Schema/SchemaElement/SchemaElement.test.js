import React from 'react';
import SchemaElement from './SchemaElement.js';
import renderer from 'react-test-renderer';

it('single-element', () => {
	// create React tree
	const tree = renderer.create(<SchemaElement name="Single_element" />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('mapped-element', () => {
	// create React tree
	const tree = renderer.create(<SchemaElement name="Mapped_element" mapped="true" />).toJSON();
	expect(tree).toMatchSnapshot();
});

it('highlighted-element', () => {
	// create React tree
	const tree = renderer
		.create(<SchemaElement name="Highlighted_element" highlighted="true" />)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
