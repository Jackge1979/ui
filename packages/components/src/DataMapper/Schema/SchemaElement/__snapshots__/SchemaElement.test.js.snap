// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`highlighted-element 1`] = `
<div
  className="schema-element highlighted undefined"
  onClick={[Function]}
  onDoubleClick={[Function]}
  role="button"
/>
`;

exports[`mapped-element 1`] = `
<div
  className="schema-element mapped undefined"
  onClick={[Function]}
  onDoubleClick={[Function]}
  role="button"
/>
`;

exports[`single-element 1`] = `
<div
  className="schema-element undefined"
  onClick={[Function]}
  onDoubleClick={[Function]}
  role="button"
/>
`;
